core/src/saros/concurrent/management/HeartbeatDispatcher.java

```diff
@@ -0,0 +1,67 @@
+package saros.concurrent.management;
+
+import java.util.Map;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import saros.activities.JupiterActivity;
+import saros.activities.SPath;
+import saros.concurrent.jupiter.internal.Jupiter;
+import saros.concurrent.jupiter.internal.text.NoOperation;
+import saros.repackaged.picocontainer.Startable;
+import saros.session.AbstractActivityProducer;
+import saros.session.ISarosSession;
+import saros.synchronize.UISynchronizer;
+import saros.util.NamedThreadFactory;
+
+/**
+ * This class sends periodically JupiterOperations to ack other clients Activities.
+ *
+ * <p>This fixes a logical memory leak, that is present in the Jupiter-implementation, if a client
+ * does not participate in the session actively (usually Saros/S). This causes other clients to hold
+ * JupiterActivities indefinitely to be able to merge them at all times. Sending 'NoOperation's
+ * fixes this issue by acking received activities though the operation's vectorTime.
+ */
+public class HeartbeatDispatcher extends AbstractActivityProducer implements Startable {
+  private ScheduledThreadPoolExecutor heartbeatScheduledExecutor;
+  private final ISarosSession sarosSession;
+  private final UISynchronizer uiSynchronizer;
+  private final Map<SPath, Jupiter> clientDocs;
+
+  public HeartbeatDispatcher(
+      ISarosSession sarosSession,
+      UISynchronizer uiSynchronizer,
+      ConcurrentDocumentClient cdClient) {
+    this.sarosSession = sarosSession;
+    this.uiSynchronizer = uiSynchronizer;
+    this.clientDocs = cdClient.getJupiterClient().getClientDocs();
```

Comment:
Holding this map of another class as an instance variable seems strange to me. It makes it hard to understand that the map is modified externally. I would prefer to hold the jupiter client instead and just request the map locally in `dispatchHeartbeats()` every time it it needed.

Even though this is probably worse performance-wise as you have to request the object multiple times (periodically), I still think it is worth it as it makes the code easier to understand.

Reply:
k